cache.go

```
@@ -0,0 +1,120 @@
+
+// -------------------------------------------------------------------------------
+type cacheError struct {
+	msg     string
+	errType errorType
+}
+
+func (ce cacheError) Error() string {
+	return ce.msg
+}
+
+type errorType string
+
+const (
+	errorTypeAlreadyExists errorType = "AlreadyExists"
+	errorTypeDoesntExist   errorType = "DoesntExist"
+)
+
+func newError(errType errorType, msg string) cacheError {
+	return cacheError{
+		msg:     msg,
+		errType: errType,
+	}
+}
+
+func IsAlreadyExistsError(err error) bool {
+	cacheErr, isCacheErr := err.(cacheError)
+	return isCacheErr && cacheErr.errType == errorTypeAlreadyExists
+}
+
+func IsDoesntExistError(err error) bool {
+	cacheErr, isCacheErr := err.(cacheError)
+	return isCacheErr && cacheErr.errType == errorTypeDoesntExist
+}
+
+// -------------------------------------------------------------------------------
+
+// -------------------------------------------------------------------------------
+type mapCache struct {
+	cacheMap map[interface{}]interface{}
+	mutex    sync.Mutex
+}
+
+// Create a new Cache object that is backed by a map
+func NewMapCache() Cache {
+	return &mapCache{
+		cacheMap: map[interface{}]interface{}{},
+	}
+}
+
+func (m *mapCache) Store(key, val interface{}, duration time.Duration) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if _, exists := m.cacheMap[key]; exists {
+		return newError(errorTypeAlreadyExists, fmt.Sprintf("key %v is already in use", key))
+	}
+
+	m.cacheMap[key] = val
+
+	if duration != 0 {
+		go func() {
+			time.Sleep(duration)
+
+			// Ignoring errors here because if the value was already removed manually we shouldn't care
+			m.Remove(key)
+		}()
+	}
+
+	return nil
+}
+
+func (m *mapCache) Remove(key interface{}) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if _, exists := m.cacheMap[key]; !exists {
+		return newError(errorTypeDoesntExist, fmt.Sprintf("key %v doesn't exist", key))
```

I don't think we should throw an error when trying to delete a key that doesn't exists, mainly because you didn't implement any function the gives the user an indication if a specific cache key exists (something like Top() in a stack). If you want to implement such function then the error will be fine